module3/src/MyList.java

```diff
@@ -0,0 +1,697 @@
+     * removes a range of elements from a list:
+     * <pre>{@code
+     *      list.subList(from, to).clear();
+     * }</pre>
+     * Similar idioms may be constructed for <tt>indexOf</tt> and
+     * <tt>lastIndexOf</tt>, and all of the algorithms in the
+     * <tt>Collections</tt> class can be applied to a subList.<p>
+     *
+     * The semantics of the list returned by this method become undefined if
+     * the backing list (i.e., this list) is <i>structurally modified</i> in
+     * any way other than via the returned list.  (Structural modifications are
+     * those that change the size of this list, or otherwise perturb it in such
+     * a fashion that iterations in progress may yield incorrect results.)
+     *
+     * @param fromIndex low endpoint (inclusive) of the subList
+     * @param toIndex high endpoint (exclusive) of the subList
+     * @return a view of the specified range within this list
+     * @throws IndexOutOfBoundsException for an illegal endpoint index value
+     *         (<tt>fromIndex &lt; 0 || toIndex &gt; size ||
+     *         fromIndex &gt; toIndex</tt>)
+     */
+    @Override
+    public List<T> subList(int fromIndex, int toIndex) {
+        //TODO Should I copy SubList subcalss from the ArrayList?
+        if(fromIndex > toIndex || fromIndex < 0 || toIndex > size-1) {
+            throw new IndexOutOfBoundsException();
+        }
+        List<T> result = new MyList<>();
+        for (int i = fromIndex; i < toIndex; i++) {
+            result.add((T) this.elementData[i]);
+        }
+        return result;
+    }
+
+    /**
+     * Retains only the elements in this list that are contained in the
+     * specified collection (optional operation).  In other words, removes
+     * from this list all of its elements that are not contained in the
+     * specified collection.
+     *
+     * @param c collection containing elements to be retained in this list
+     * @return <tt>true</tt> if this list changed as a result of the call
+     * @throws UnsupportedOperationException if the <tt>retainAll</tt> operation
+     *         is not supported by this list
+     * @throws ClassCastException if the class of an element of this list
+     *         is incompatible with the specified collection
+     * (<a href="Collection.html#optional-restrictions">optional</a>)
+     * @throws NullPointerException if this list contains a null element and the
+     *         specified collection does not permit null elements
+     *         (<a href="Collection.html#optional-restrictions">optional</a>),
+     *         or if the specified collection is null
+     * @see #remove(Object)
+     * @see #contains(Object)
+     */
+    @Override
+    public boolean retainAll(Collection<?> c) {
+        if (c.size() == 0) {
+            return false;
+        }
+        boolean wasChanged = false;
+        int i = 0;
+        while (i < size) {
+            if (!c.contains(elementData[i])) {
```

Comment:
do you really need this `!` not?

Reply:
I do not.
Reworked this IF ELSE to work without the `!`:
```
            if (c.contains(elementData[i])) {
                i++;
            } else {
                this.remove(i);
                wasChanged = true;
            }
```